lib/reverseproxy/reverse_proxy.go

```diff
@@ -0,0 +1,130 @@
+package reverseproxy
+
+import (
+	"flag"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+)
+
+const (
+	targetURL     = "vm-target-url"
+	forwardHeader = "X-Forwarded-For"
+)
+
+var (
+	maxIdleConnsPerBackend = flag.Int("maxIdleConnsPerBackend", 100, "The maximum number of idle connections vmauth can open per each backend host")
+)
+
+var bbPool bytesutil.ByteBufferPool
+
+// ReverseProxy represents simple revers proxy based on http.Client
+type ReverseProxy struct {
+	client *http.Client
+}
+
+// New initialize revers proxy
+func New() *ReverseProxy {
+	tr := prepareTransport()
+	return &ReverseProxy{
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+}
+
+// ServeHTTP serve http requests
+func (rr *ReverseProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rr.handle(w, r)
+}
+
+// handle works with income request, update it to outcome and copy response to the requester
+func (rr *ReverseProxy) handle(w http.ResponseWriter, r *http.Request) {
+	bb := bbPool.Get()
+	defer func() { bbPool.Put(bb) }()
+
+	tURL, err := getTargetURL(r)
+	if err != nil {
+		logger.Panicf("BUG: unexpected error when parsing targetURL=%q: %s", targetURL, err)
+	}
+
+	ctx := r.Context()
+	proxyReq := r.Clone(ctx)
+
+	proxyReq.URL = tURL
+	if r.ContentLength == 0 {
+		proxyReq.Body = nil // Issue 16036: nil Body for http.Transport retries
+	}
+	if proxyReq.Body != nil {
+		defer func() { _ = proxyReq.Body.Close() }()
+	}
+
+	if err := updateXForwardHeader(r.RemoteAddr, proxyReq.Header); err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+
+	resp, err := rr.client.Transport.RoundTrip(proxyReq)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	for k, vv := range resp.Header {
+		for _, v := range vv {
+			w.Header().Set(k, v)
+		}
+	}
+
+	w.WriteHeader(resp.StatusCode)
+	_, err = io.CopyBuffer(w, resp.Body, bb.B)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
```

Comment:
move logic to separate function that return err, handle this error in origin function

Reply:
updated